design/0003-isolated-integration-tests.md

```diff
@@ -0,0 +1,324 @@
+- Have a more discoverable UX for running all or one test
+- Less side effects when running the tests (and less footguns)
+- Easier to extend for more complex integration tests
+
+The two main components that the following subsections talk about are the code
+component and the script component.
+
+The code component is comprised of the Go files in the `test/` directory and
+have direct interaction with `go test`. The Go test files located here are
+responsible for testing specific features of Linkerd such as service profiles,
+tap, edges, and more--they are not the actual integration tests. It is
+important to note that these tests assume the cluster state is already set,
+and it should stay that way. This allows developers to run `go test` and
+specify the cluster context, namespace, and other knobs that may be important
+for properly testing certain cases.
+
+The script component is comprised of the `test-run` and `_test-run.sh`
+scripts. These are the actual integration tests. Each integration test is
+comprised of cluster setup and some combination of the test files from the
+code component.
+
+The separation here is important to note for this change: the integration
+tests are contained within scripts and responsible for managing the cluster
+state and running some combination of Go test files.
+
+Therefore, these changes will be mostly script based. There is some possible
+tooling that will allow managing cluster state through code (discussed in `#
+Future  possibilities`) but the ecosystem is currently not in a state where we
+can avoid breaking out to the shell at some point for configuring something
+about the environment.
+
+### UX
+
+Running the integration tests will be done through new scripts that are
+responsible for the setup and teardown of their testing environments.
+
+This is different from the current situation in a few ways. First, there is
+only one self-contained script that runs the integration test suit; it runs
+all the tests and does not provide a way to select specific ones. If one does
+wish to run a specific test, they must export the internal test script into
+their shell environment.
+
+Second, the environment setup check must be run before each integration test.
+This is automatic in the self-contained script, but is a required manual step
+for running a specific one.
+
+It will be important to document how developers can run all or one of the
+tests, so providing clear instructions in the Test markdown file, as well as a
+`--help` flag for the scripts will be included.
+
+Before: Running all tests:
+
+```
+$ bin/test-run $PWD/target/cli/darwin/linkerd
+```
+
+Before: Running specific test:
+
+```
+$ . bin/_test-run.sh
+$ init_test_run $PWD/target/cli/darwin/linkerd
+$ deep_integration_tests
+```
+
+After: Running all tests:
+
+```
+$ bin/tests $PWD/target/cli/darwin/linkerd
+```
+
+After: Running specific test:
+
+```
+$ bin/deep-tests $PWD/target/cli/darwin/linkerd
```

Comment:
Also, I'm confused by the concept of `bin/deep-tests`... What's a "deep" test? What other kinds of tests are there? As a user, how do I know which script I want to run and why?

Reply:
This is just the current [deep_integration_test](https://github.com/linkerd/linkerd2/blob/master/bin/_test-run.sh#L87-L94) that runs all the Go test files with no additional arguments.

There are integration tests such as uninstall or external issuer that would pass additional flags to the Go tests files, such as `--external-issuer=true`.

With the integration tests being separate scripts, the idea is that you would know what you want to run based off the name: `bin/external-issuer-test`, `bin/uninstall-test`, `bin/deep-test` (which maybe could be renamed to `all-test`/`all-default-test`)

Reply:
What is it about these tests that vary, specifically?

Do they handle cluster setup/teardown different? Do they all end up invoking `go test`?

Also, we should consider a naming hierarchy that is tab-complete/sort friendly like `bin/test-all`.

Is there a way for this to be structured as, for instance `bin/test-harness ./path/to/issuer/tests`?

How does a developer understanding the mapping between the various test scripts and the actual source-code? Are the tests split into separate directories? If so, should we just keep a runner script in each test directory?

I don't really have enough context about the current state or how it's laid out to provide a lot in the way of guidance here, I guess. But, that probably makes me a decent advocate for an uninitiated contributor... Is the current setup documented? If I wanted to start adding a new type of tests, is there guidance in how I should approach it?

Reply:
> Do they handle cluster setup/teardown different? Do they all end up invoking go test

They all end up invoking `go test` and passing in different flags specific to the test. For example, testing external issuer integration test will:
1. Run `install_test.go` (which is a test within itself for checking a clean install with the flags passed) with `--external-issuer=true` so that the setup installs a cluster with external issuer
2. Run `externalissuer/...` tests
3. Cleanup the namespace

The external issuer integration test takes care of this setup and teardown, but allows a developer to set some other specific state and run `go test github.com/linkerd/linkerd2/test/externalissuer/... [flags]`

> Also, we should consider a naming hierarchy that is tab-complete/sort friendly like bin/test-all

Sure that would make sense to do!

Reply:
> Is there a way for this to be structured as, for instance bin/test-harness ./path/to/issuer/tests?

This is what `bin/test-external-issuer` would be. It setups the states, runs the test specified by the external issuer path, passes in the flags required (context, namespace, CLI) and then performs cleanup.